code/datums/components/customizable_reagent_holder.dm

```diff
@@ -0,0 +1,245 @@
+			var/obj/item/I = ingr
+			add_ingredient(I)
+			handle_fill(I)
+
+
+/datum/component/customizable_reagent_holder/Destroy(force, silent)
+	QDEL_NULL(top_overlay)
+	return ..()
+
+
+/datum/component/customizable_reagent_holder/RegisterWithParent()
+	. = ..()
+	RegisterSignal(parent, COMSIG_PARENT_ATTACKBY, .proc/customizable_attack)
+	RegisterSignal(parent, COMSIG_PARENT_EXAMINE, .proc/on_examine)
+	RegisterSignal(parent, COMSIG_ATOM_PROCESSED, .proc/on_processed)
+
+
+/datum/component/customizable_reagent_holder/UnregisterFromParent()
+	. = ..()
+	UnregisterSignal(parent, list(
+			COMSIG_PARENT_ATTACKBY,
+			COMSIG_PARENT_EXAMINE,
+			COMSIG_ATOM_PROCESSED
+		)
+	)
+
+
+/datum/component/customizable_reagent_holder/PostTransfer()
+	if(!isatom(parent))
+		return COMPONENT_INCOMPATIBLE
+	var/atom/P = parent
+	if (!P.reagents)
+		return COMPONENT_INCOMPATIBLE
+
+///Handles when the customizable food is examined.
+/datum/component/customizable_reagent_holder/proc/on_examine(atom/A, mob/user, list/examine_list)
+	SIGNAL_HANDLER
+
+	var/atom/P = parent
+	var/ingredients_listed = ""
+	if (LAZYLEN(ingredients))
+		for (var/i in 1 to ingredients.len)
+			var/obj/item/I = ingredients[i]
+			var/ending = ", "
+			switch(length(ingredients))
+				if (2)
+					if (i == 1)
+						ending = " and "
+				if (3 to INFINITY)
+					if (i == ingredients.len - 1)
+						ending = ", and "
+			ingredients_listed += "\a [I.name][ending]"
+	examine_list += "It contains [LAZYLEN(ingredients) ? "[ingredients_listed]" : " no ingredients, "]making a [custom_adjective()]-sized [initial(P.name)]."
+
+
+///Handles when the customizable food is attacked by something.
+/datum/component/customizable_reagent_holder/proc/customizable_attack(datum/source, obj/item/I, mob/M, silent = FALSE, force = FALSE)
+	SIGNAL_HANDLER
+
+	var/valid_ingredient = TRUE
+
+	switch (ingredient_type)
+		if (CUSTOM_INGREDIENT_TYPE_EDIBLE)
+			valid_ingredient = IS_EDIBLE(I)
+
+	// only accept valid ingredients
+	if (!valid_ingredient)
+		to_chat(M, "<span class='warning'>[I] doesn't belong on [parent]!</span>")
+		return
+
+	if (LAZYLEN(ingredients) >= max_ingredients)
+		to_chat(M, "<span class='warning'>[parent] is too full for any more ingredients!</span>")
+		return
+
+	var/atom/P = parent
+	if(!M.transferItemToLoc(I, P))
+		return
+	if (replacement)
+		var/atom/R = new replacement(P.loc)
+		I.forceMove(R)
+		replacement = null
+		RemoveComponent()
+		R.TakeComponent(src)
+		qdel(P)
+	handle_reagents(I)
+	add_ingredient(I)
+	handle_fill(I)
+
+
+///Handles the icon update for a new ingredient.
+/datum/component/customizable_reagent_holder/proc/handle_fill(obj/item/I)
+	if (fill_type == CUSTOM_INGREDIENT_ICON_NOCHANGE)
+		//don't bother doing the icon procs
+		return
+	var/atom/P = parent
+	var/mutable_appearance/filling = mutable_appearance(P.icon, "[initial(P.icon_state)]_filling")
+	// get average color
+	var/icon/ico = new(I.icon, I.icon_state)
+	ico.Scale(1, 1)
```

Comment:
Would rather this just be `icon`, we're not losing anything with one more letter.